src/main/java/org/craftercms/engine/service/context/SiteContextManager.java

```diff
@@ -197,6 +255,139 @@ public void createContexts(boolean concurrent) {
+                                // Only process if the hash is different from the last processed hash value,
+                                // or the event hash is null in the case of DELETE
+                                // This means the modified time has been updated from last processed event or the file is deleted
+                                // This prevents multiple events for batch files change with same modified date such as from a git pull
+                                String lastProcessedHash = directoryWatcherLastProcessedHash.get(siteName);
+                                if (lastProcessedHash == null || event.hash() == null || !lastProcessedHash.equals(hashValue)) {
+                                    Lock siteLock = siteLockFactory.getLock(siteName);
+                                    siteLock.lock();
+                                    try {
+                                        if (event.hash() != null) {
+                                            directoryWatcherLastProcessedHash.put(siteName, hashValue);
+                                        }
+                                        AtomicInteger counter = directoryWatcherCounter.get(siteName);
+                                        if (counter == null) {
+                                            counter = new AtomicInteger(0);
+                                        }
+                                        counter.getAndIncrement();
+                                        directoryWatcherCounter.put(siteName, counter);
+                                    } finally {
+                                        siteLock.unlock();
+                                    }
+                                } else {
+                                    logger.info("File watch for hash '{}' has already processed. No action required.", hashValue);
+                                }
+                                break;
+                            }
+                            default:
+                                logger.info("File watcher unhandled event kind: '{}'. File affected: '{}'.", event.eventType(), event.path());
+                        }
+                    })
+                    .build();
+
+            // Remove old watcher before register a new one
+            if (directoryWatcherRegistry.get(siteName) != null) {
+                Lock siteLock = siteLockFactory.getLock(siteName);
+                siteLock.lock();
+                try {
+                    DirectoryWatcher oldWatcher = directoryWatcherRegistry.remove(siteName);
+                    oldWatcher.close();
+                } finally {
+                    siteLock.unlock();
+                }
+            }
+            directoryWatcherRegistry.put(siteName, watcher);
+            watcher.watchAsync();
+        } catch (Exception e) {
+            logger.error("Error while creating watcher for site: " + siteName, e);
```

Comment:
Switch to a String.format instead of string concatenation.

Reply:
Updated. Thank you!